File: webrtc/src/main/java/dev/onvoid/webrtc/RTCRtpSynchronizationSource.java
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.onvoid.webrtc;

/**
 * Contains information about a given synchronization source (SSRC).
 *
 * @author Alex Andres
 */
public class RTCRtpSynchronizationSource extends RTCRtpContributingSource {

	/**
	 * Only present for audio receivers. Whether the last RTP packet, delivered
	 * from this source, contains voice activity (true) or not (false). If the
	 * RFC 6464 extension header was not present, or if the peer has signaled
	 * that it is not using the V bit by setting the "vad" extension attribute
	 * to "off".
	 */
	public final boolean voiceActivityFlag;


	/**
	 * Creates an instance of RTCRtpSynchronizationSource with the specified
	 * synchronization source information.
	 *
	 * @param timestamp         The most recent time a frame from an RTP
	 *                          packet.
	 * @param source            The CSRC or SSRC identifier of the contributing
	 *                          or synchronization source.
	 * @param audioLevel        The sound pressure level. Only present for audio
	 *                          receivers.
	 * @param rtpTimestamp      The last RTP timestamp.
	 * @param voiceActivityFlag Whether the last RTP packet contains voice
	 *                          activity (true) or not (false).
	 */
	protected RTCRtpSynchronizationSource(long timestamp, long source,
			double audioLevel, long rtpTimestamp, boolean voiceActivityFlag) {
		super(timestamp, source, audioLevel, rtpTimestamp);

		this.voiceActivityFlag = voiceActivityFlag;
	}
}

File: webrtc-jni/src/main/cpp/src/api/RTCRtpSynchronizationSource.cpp
/*
 * Copyright 2019 Alex Andres
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "api/RTCRtpSynchronizationSource.h"
#include "api/RTCRtpContributingSource.h"
#include "JavaObject.h"
#include "JavaRef.h"
#include "JavaUtils.h"
#include "JNI_WebRTC.h"

namespace jni
{
	namespace RTCRtpSynchronizationSource
	{
		JavaLocalRef<jobject> toJava(JNIEnv * env, const webrtc::RtpSource & source)
		{
			const auto javaClass = JavaClasses::get<JavaRTCRtpSynchronizationSourceClass>(env);

			jlong timestamp = static_cast<jlong>(source.timestamp_ms());
			jlong sourceId = static_cast<jlong>(source.source_id());
			jdouble audioLevel = static_cast<jdouble>(source.audio_level().value_or(0));
			jlong rtpTimestamp = static_cast<jlong>(source.rtp_timestamp());
			jboolean voiceActivityFlag = false;

			jobject object = env->NewObject(javaClass->cls, javaClass->ctor,
				timestamp, sourceId, audioLevel, rtpTimestamp, voiceActivityFlag);

			return JavaLocalRef<jobject>(env, object);
		}

		webrtc::RtpSource toNative(JNIEnv * env, const JavaRef<jobject> & source)
		{
			const auto javaClass = JavaClasses::get<JavaRTCRtpSynchronizationSourceClass>(env);
			const auto parentClass = JavaClasses::get<RTCRtpContributingSource::JavaRTCRtpContributingSourceClass>(env);

			JavaObject obj(env, source);

			return webrtc::RtpSource(
				static_cast<int64_t>(obj.getLong(parentClass->timestamp)),
				static_cast<uint32_t>(obj.getLong(parentClass->source)),
				webrtc::RtpSourceType::SSRC,
				static_cast<uint8_t>(obj.getDouble(parentClass->audioLevel)),
				static_cast<uint32_t>(obj.getLong(parentClass->rtpTimestamp))
			);
		}

		JavaRTCRtpSynchronizationSourceClass::JavaRTCRtpSynchronizationSourceClass(JNIEnv * env)
		{
			cls = FindClass(env, PKG"RTCRtpSynchronizationSource");

			ctor = GetMethod(env, cls, "<init>", "(JJDJZ)V");

			voiceActivityFlag = GetFieldID(env, cls, "voiceActivityFlag", "Z");
		}
	}
}
